### DISCUSSION TITLE
v7.2.0 — Full Release

### DISCUSSION THREAD

**Discussion by apprenticeharper** [apprenticeharper/DeDRM_tools#1655] [General]
This release incorporates numerous small bug fixes contributed through pull requests to the GitHub repository. Thank you to all contributors.

It also now includes code that should handle the latest KFX encryption used by the latest Kindle for PC and Kindle for Mac. How long this code will work is anyone's guess. It may be that Amazon will update Kindle for PC/Mac very quickly. In which case, the existing work-arounds to get KF8 format should still work.

I am very grateful to those dedicated developers who spend their free time reverse-engineering the DRM code.

<hr /><em>This discussion was created from the release <a href='https://github.com/apprenticeharper/DeDRM_tools/releases/tag/v7.2.0'>v7.2.0 — Full Release</a>.</em>

**Comment by pdurrant** [comment]
I get:

 File "calibre_plugins.dedrm.kfxdedrm", line 85
    traceback.print_exc()
                        ^
TabError: inconsistent use of tabs and spaces in indentation

When trying to import a book from Kindle for Mac. 7.1.0 was working fine for me.

**Comment by deirdrew** [comment]
I just had to revert to 7.1.0 as it failed on one of two Amazon books I just got using Kindle for PC 1.17.1. As soon as I reverted it was successful.

**Reply by pdurrant** [reply]
It looks like there's a 7.2.1 already.

**Reply by pdurrant** [reply]
It works for me for AZW3

**Reply by deirdrew** [reply]
Just saw that (sorry, was testing 7.2.0 on a different PC that has Kindle for PC latest version and didn't see the new email). Yes, 7.2.1 was able to handle the book 7.20 could not. :)

**Comment by TheRealArtfulDodger** [comment]
I just saw that dedrm was working with Calibre version 5. So, I tried to install it and I received the message I've attached. It's likely I'm doing something Wrong but...
I'm in Windows 10 v. 2004 x64 and using Calibre Portable 5.17.
Sparky

**Reply by TheRealArtfulDodger** [reply]
I see the attachment didn't work - I shall try again, else I shall type it in! I think this time it worked!
![Plugin Error](https://user-images.githubusercontent.com/72087375/123275702-771a6e00-d4fc-11eb-9c94-2ae411072494.png)
Sparky

**Reply by ElleKayEm** [reply]
Unzip DeDRM_tools_7.2.1.zip.  Load DeDRM_plugin.zip.

**Reply by TheRealArtfulDodger** [reply]
Hello, ElleKayEm:
Well, now I feel like a real fool! From this little demonstration, you'd never know I've only been working with computers since 1962!
Thank you - As I said, I just knew it was something I was doing wrong.
Again, thank you very much for your help!
Sparky

**Comment by gailong** [comment]
I upgrade my widnows to windows 11 , and I found I can't import the patch

calibre, version 5.30.0
錯誤： 未處理的例外: <b>InvalidPlugin</b>:The plugin in 'C:\\Users\\gnoli\\Downloads\\DeDRM_tools_7.2.1.zip' is invalid. It does not contain a top-level __init__.py file

calibre 5.30  embedded-python: True is64bit: False
Windows-10-10.0.22000 Windows ('32bit', 'WindowsPE')
32bit process running on 64bit windows
('Windows', '10', '10.0.22000')
Python 3.8.5
Windows: ('10', '10.0.22000', '', 'Multiprocessor Free')
Interface language: zh_TW
Traceback (most recent call last):
  File "calibre\gui2\preferences\plugins.py", line 327, in add_plugin
  File "calibre\customize\ui.py", line 476, in add_plugin
  File "calibre\customize\ui.py", line 64, in load_plugin
  File "calibre\customize\zipplugin.py", line 284, in load
  File "calibre\customize\zipplugin.py", line 379, in _locate_code
calibre.customize.InvalidPlugin: The plugin in 'C:\\Users\\gnoli\\Downloads\\DeDRM_tools_7.2.1.zip' is invalid. It does not contain a top-level __init__.py file

**Reply by ElleKayEm** [reply]
Unzip DeDRM_tools_7.2.1.zip and load DeDRM_plugin.zip.

**Comment by Baghelnaveen134** [comment]
calibre, version 5.9.0
ERROR: Conversion failed: <p><b>Failed to convert: The Axe Falls<p>This is an Amazon KFX book. It cannot be processed. See <a href="https://www.mobileread.com/forums/showthread.php?t=283371">here</a> for information on how to handle KFX books.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "runpy.py", line 194, in _run_module_as_main
  File "runpy.py", line 87, in _run_code
  File "site.py", line 82, in <module>
  File "site.py", line 77, in main
  File "site.py", line 49, in run_entry_point
  File "calibre\utils\ipc\worker.py", line 216, in main
  File "calibre\gui2\convert\gui_conversion.py", line 41, in gui_convert_override
  File "calibre\gui2\convert\gui_conversion.py", line 28, in gui_convert
  File "calibre\ebooks\conversion\plumber.py", line 1108, in run
  File "calibre\customize\conversion.py", line 245, in __call__
  File "calibre\ebooks\conversion\plugins\mobi_input.py", line 36, in convert
  File "calibre\ebooks\mobi\reader\mobi6.py", line 85, in __init__
calibre.ebooks.mobi.reader.mobi6.KFXError: This is an Amazon KFX book. It cannot be processed. See <a href="https://www.mobileread.com/forums/showthread.php?t=283371">here</a> for information on how to handle KFX books.

Convert book 1 of 1 (The Axe Falls)
DeDRM v7.2.1: Trying to decrypt awezzh7x.azw
Using Library AlfCrypto DLL/DYLIB/SO
Using Library AlfCrypto DLL/DYLIB/SO
DeDRM v7.2.1: Failed to decrypt with error: The .kfx DRMION file cannot be decrypted by itself. A .kfx-zip archive containing a DRM voucher is required.
DeDRM v7.2.1: Looking for new default Kindle Key after 0.2 seconds
searching for kinfoFiles in C:\Users\Naveen\AppData\Local
Found K4PC 1.25+ kinf2018 file: C:\Users\Naveen\AppData\Local\Amazon\Kindle\storage\.kinf2018
Decrypted key file using IDString '2150616300' and UserName 'Naveen'
DeDRM v7.2.1: Found 1 new key
DeDRM v7.2.1: Ultimately failed to decrypt after 8.3 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md
Conversion options changed from defaults:
  verbose: 2
  read_metadata_from_opf: 'C:\\Users\\Naveen\\AppData\\Local\\Temp\\calibre_0sohc96a\\_0g28jj0.opf'
  output_profile: 'generic_eink'
Resolved conversion options
calibre version: 5.9.0
{'asciiize': False,
 'author_sort': None,
 'authors': None,
 'base_font_size': 0.0,
 'book_producer': None,
 'change_justification': 'original',
 'chapter': "//*[((name()='h1' or name()='h2') and re:test(., "
            "'\\s*((chapter|book|section|part)\\s+)|((prolog|prologue|epilogue)(\\s+|$))', "
            "'i')) or @class = 'chapter']",
 'chapter_mark': 'pagebreak',
 'comments': None,
 'cover': None,
 'debug_pipeline': None,
 'dehyphenate': True,
 'delete_blank_paragraphs': True,
 'disable_font_rescaling': False,
 'dont_split_on_page_breaks': False,
 'duplicate_links_in_toc': False,
 'embed_all_fonts': False,
 'embed_font_family': None,
 'enable_heuristics': False,
 'epub_flatten': False,
 'epub_inline_toc': False,
 'epub_toc_at_end': False,
 'epub_version': '2',
 'expand_css': False,
 'extra_css': None,
 'extract_to': None,
 'filter_css': '',
 'fix_indents': True,
 'flow_size': 260,
 'font_size_mapping': None,
 'format_scene_breaks': True,
 'html_unwrap_factor': 0.4,
 'input_encoding': None,
 'input_profile': <calibre.customize.profiles.InputProfile object at 0x08D4ED90>,
 'insert_blank_line': False,
 'insert_blank_line_size': 0.5,
 'insert_metadata': False,
 'isbn': None,
 'italicize_common_cases': True,
 'keep_ligatures': False,
 'language': None,
 'level1_toc': None,
 'level2_toc': None,
 'level3_toc': None,
 'line_height': 0.0,
 'linearize_tables': False,
 'margin_bottom': 5.0,
 'margin_left': 5.0,
 'margin_right': 5.0,
 'margin_top': 5.0,
 'markup_chapter_headings': True,
 'max_toc_links': 50,
 'minimum_line_height': 120.0,
 'no_chapters_in_toc': False,
 'no_default_epub_cover': False,
 'no_inline_navbars': False,
 'no_svg_cover': False,
 'output_profile': <calibre.customize.profiles.GenericEink object at 0x08D4EF10>,
 'page_breaks_before': "//*[name()='h1' or name()='h2']",
 'prefer_metadata_cover': False,
 'preserve_cover_aspect_ratio': False,
 'pretty_print': True,
 'pubdate': None,
 'publisher': None,
 'rating': None,
 'read_metadata_from_opf': 'C:\\Users\\Naveen\\AppData\\Local\\Temp\\calibre_0sohc96a\\_0g28jj0.opf',
 'remove_fake_margins': True,
 'remove_first_image': False,
 'remove_paragraph_spacing': False,
 'remove_paragraph_spacing_indent_size': 1.5,
 'renumber_headings': True,
 'replace_scene_breaks': '',
 'search_replace': '[]',
 'series': None,
 'series_index': None,
 'smarten_punctuation': False,
 'sr1_replace': None,
 'sr1_search': None,
 'sr2_replace': None,
 'sr2_search': None,
 'sr3_replace': None,
 'sr3_search': None,
 'start_reading_at': None,
 'subset_embedded_fonts': False,
 'tags': None,
 'timestamp': None,
 'title': None,
 'title_sort': None,
 'toc_filter': None,
 'toc_threshold': 6,
 'toc_title': None,
 'transform_css_rules': '[]',
 'unsmarten_punctuation': False,
 'unwrap_lines': True,
 'use_auto_toc': False,
 'verbose': 2}
Running file type plugin DeDRM failed with traceback:
Traceback (most recent call last):
  File "calibre_plugins.dedrm.__init__", line 540, in KindleMobiDecrypt
  File "calibre_plugins.dedrm.k4mobidedrm", line 208, in GetDecryptedBook
calibre_plugins.dedrm.k4mobidedrm.DrmException: The .kfx DRMION file cannot be decrypted by itself. A .kfx-zip archive containing a DRM voucher is required.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "calibre\customize\ui.py", line 172, in _run_filetype_plugins
  File "calibre_plugins.dedrm.__init__", line 633, in run
  File "calibre_plugins.dedrm.__init__", line 583, in KindleMobiDecrypt
calibre_plugins.dedrm.DeDRMError: DeDRM v7.2.1: Ultimately failed to decrypt after 8.3 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md
DeDRM v7.2.1: Trying to decrypt awezzh7x.azw
DeDRM v7.2.1: Failed to decrypt with error: The .kfx DRMION file cannot be decrypted by itself. A .kfx-zip archive containing a DRM voucher is required.
DeDRM v7.2.1: Looking for new default Kindle Key after 0.0 seconds
searching for kinfoFiles in C:\Users\Naveen\AppData\Local
Found K4PC 1.25+ kinf2018 file: C:\Users\Naveen\AppData\Local\Amazon\Kindle\storage\.kinf2018
Decrypted key file using IDString '2150616300' and UserName 'Naveen'
DeDRM v7.2.1: Found 1 new key
DeDRM v7.2.1: Ultimately failed to decrypt after 8.1 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md
Running file type plugin DeDRM failed with traceback:
Traceback (most recent call last):
  File "calibre_plugins.dedrm.__init__", line 540, in KindleMobiDecrypt
  File "calibre_plugins.dedrm.k4mobidedrm", line 208, in GetDecryptedBook
calibre_plugins.dedrm.k4mobidedrm.DrmException: The .kfx DRMION file cannot be decrypted by itself. A .kfx-zip archive containing a DRM voucher is required.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "calibre\customize\ui.py", line 172, in _run_filetype_plugins
  File "calibre_plugins.dedrm.__init__", line 633, in run
  File "calibre_plugins.dedrm.__init__", line 583, in KindleMobiDecrypt
calibre_plugins.dedrm.DeDRMError: DeDRM v7.2.1: Ultimately failed to decrypt after 8.1 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md
InputFormatPlugin: MOBI Input running
on C:\Users\Naveen\AppData\Local\Temp\calibre_0sohc96a\awezzh7x.azw
Traceback (most recent call last):
  File "calibre\ebooks\conversion\plugins\mobi_input.py", line 30, in convert
  File "calibre\ebooks\mobi\reader\mobi6.py", line 85, in __init__
calibre.ebooks.mobi.reader.mobi6.KFXError: This is an Amazon KFX book. It cannot be processed. See https://www.mobileread.com/forums/showthread.php?t=283371 for information on how to handle KFX books.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "runpy.py", line 194, in _run_module_as_main
  File "runpy.py", line 87, in _run_code
  File "site.py", line 82, in <module>
  File "site.py", line 77, in main
  File "site.py", line 49, in run_entry_point
  File "calibre\utils\ipc\worker.py", line 216, in main
  File "calibre\gui2\convert\gui_conversion.py", line 41, in gui_convert_override
  File "calibre\gui2\convert\gui_conversion.py", line 28, in gui_convert
  File "calibre\ebooks\conversion\plumber.py", line 1108, in run
  File "calibre\customize\conversion.py", line 245, in __call__
  File "calibre\ebooks\conversion\plugins\mobi_input.py", line 36, in convert
  File "calibre\ebooks\mobi\reader\mobi6.py", line 85, in __init__
calibre.ebooks.mobi.reader.mobi6.KFXError: This is an Amazon KFX book. It cannot be processed. See https://www.mobileread.com/forums/showthread.php?t=283371 for information on how to handle KFX books.

**Reply by ElleKayEm** [reply]
For kfx files, use DeDRM 10.0.9: https://github.com/noDRM/DeDRM_tools/releases/tag/v10.0.9

And KFX Input: https://www.mobileread.com/forums/showthread.php?t=291290
